### DISCUSSION TITLE
Recommended way for running github actions

### DISCUSSION THREAD

**Discussion by TomSaw** [modm-io/modm#1167] [Q&A] [answered]
Before pushing changes to an active PR, how do you pass the tests?

I can run tests in /tests/* from the terminal, ok, but i also want to cover these github actions to minimize notification noise for others.

Found these attempts:

1. Work with a parallel stagging branch
2. Run github actions locally using a tool like [act](https://medium.com/@debasishkumardas5/running-github-actions-locally-a-complete-guide-for-windows-mac-and-linux-users-34c45999c7cd)
3. ... ?

What's your choice or recommendation?

**Comment by salkinium** [comment, marked as answer]
The GitHub actions are all just invoking Python scripts:

- Examples are compiled via the `tools/scripts/examples_compile.py {folders}` script.
- The HAL tests are compile via the `test/all/run_all.py` script.
- Unittests have [their own Makefile](https://github.com/modm-io/modm/blob/develop/.github/workflows/linux.yml#L65-L83).
- Some other small tests are mostly [also Python scripts in the `tools/scripts` folder](https://github.com/modm-io/modm/blob/develop/.github/workflows/linux.yml#L37-L60).

I personally just check the Python scripts and then check what fails, then reproduce all bugs locally and fix them, then push a group of commits again. I would never compile all examples and HAL tests locally, as it can easily generate dozens of GBs of build artifacts altogether and it's really slow even natively.

(Don't worry about notification noise, I disabled CI runs anyways, I don't find those notifications very useful.)

**Comment by TomSaw** [comment]
Thank you salk', just the checklist i needed!

**Comment by TomSaw** [comment]
I've run some HAL tests locally amd some targets failed, f.e. atmega324pv-* , why is that? And why does the CI not worry about it?
```
cd /test/all; python run_all.py atmega324pv
Using 8 parallel jobs for 6 devices
...
```
Spawns dozens like:
```
error: 'SPR0' was not declared in this scope; did you mean 'SPSR0'?
```
```
...

SUCCESS    0
FAIL       6
------------------
Total      6
```

**Reply by salkinium** [reply]
There is [a list of device prefixes that are ignored (everywhere)](https://github.com/modm-io/modm/blob/develop/test/all/ignored.txt), often due to toolchain issues unrelated to modm, or simply because their not supported yet (like the dual-core STM32H7s).

We have no support for the new AVRs (produced by Microchip, not Atmel), since their peripherals are often very different from the old ones. It's a bit of a shame, since they are slightly better, but then again absolutely no match for a Cortex-M.
